### ISSUE TITLE
moreh: offline kernel compilation

### ISSUE THREAD

**Issue by pgkeller** [tenstorrent/tt-metal#11006]
Create a new runtime API, eg: CreateKernelFromBinary
This skips runtime compilation using an offline compiled binary

Create a mechanism for compiling kernel binaries offline.  Since our binaries are heavily customized via defines/common runtime args, need to be sure kernel hashes match at runtime w/ what was generated at offline compile time.

Due date: Q4 (labelling P2 for now as work on this is deferred for now)

**Comment by aliuTT** [comment]
With an offline kernel compile, we can build device dispatch kernels (i.e. `compile_command_queue_programs()`) offline and remove the use of persistent cache in `device.cpp`:
```
detail::EnablePersistentKernelCache();
this->compile_command_queue_programs();
detail::DisablePersistentKernelCache();
```

**Comment by razorback3** [comment]
As the due date mentioned in the issue description is out of date, would you update the schedule for this issue?

**Comment by pgkeller** [comment]
This is being worked on.  We should have offline firmware build and internal kernel builds (eg, dispatch kernels) in the next week.  Generic kernel builds are more complicated.  We'll have a first cut version in the next three weeks (first week of Feb).  There may be some iteration there as we get feedback.

**Comment by kmabeeTT** [comment]
Seems related to / partially inspired by https://github.com/tenstorrent/tt-metal/issues/11010 (both tickets opened same date), right?  Curious about using this on my side if yes.

**Comment by razorback3** [comment]
> Seems related to / partially inspired by [#11010](https://github.com/tenstorrent/tt-metal/issues/11010) (both tickets opened same date), right? Interested about using this on my side if yes, subcribing to this ticket.

Yes, they look identical.

**Comment by ihnpark551** [comment]
Hi @pgkeller, could you share any updates on your side for this ticket?
This item needs to be ready by the time we begin packaging, 
therefore I set this ticket's priority to P1.

**Comment by pgkeller** [comment]
> This item needs to be ready by the time we begin packaging,
Hi @ihnpark551 what date is that approximately?

**Comment by ihnpark551** [comment]
@pgkeller 
It will align with our plan if it's ready by the end of September. 
Do you think that'll be possible on your side?

**Comment by zzigler-tt** [comment]
@pgkeller reminder to scope the work and confirm timelines for completion - est mid Oct??

**Comment by pgkeller** [comment]
yes, we'll target end of September to give us a week or so of slip.  The change we need to make to meet the minimum requirement we discussed is not too complex, so we should be able to hit the earlier side of the schedule

**Comment by ihnpark551** [comment]
Hi @pgkeller , how have you been?
about this ticket, do you have any updates on your end?

**Comment by pgkeller** [comment]
yes, plan is done, work has begun, should have a first cut ~next week

**Comment by ihnpark551** [comment]
@pgkeller Understood, can I regard the completion date of this ticket as 09/30, PST?

**Comment by pgkeller** [comment]
there will be something for you to try this week.  we had to make some compromises since the build system rework is further out, but I think it'll meet your needs in the short/medium term

**Comment by ihnpark551** [comment]
@pgkeller Thanks for your sharing, then would wait for your updates on this week!

**Comment by kstevensTT** [comment]
The PR for the Moreh binary flow update is available:
PR: https://github.com/tenstorrent/tt-metal/pull/29806
Branch: https://github.com/tenstorrent/tt-metal/tree/prybicki/aot-compilation-poc

The flow to use it is as follows.

Using `CreateKernelFromBinary`:
1. Run the program normally to generate the kernel binaries
2. Copy the generated kernel directories under the root directory (e.g. from ~/.cache/tt-metal-cache/<hash>/<build_key>/kernels/<kernel_dirs>). You can copy only the directories & binaries you want to use, but you must preserve the full path under the root.
3. Call `tt::tt_metal::experimental::ComputeKernelOriginalPathHash` on the path passed into `CreateKernel` to generate the hash of the original path.
4. Update program to call `tt_metal::experimental::SetKernelBinaryPathPrefix`. Argument `binary_path_prefix` is the location the kernel directories were copied to in step 2.
5. Call `tt:tt_metal::experimental::CreateKernelFromBinary`.
	1. `kernel_name` is the name of the kernel.
	2. `original_path_or_hash` is the hash generated by step 3.

Using `KernelDescriptor` (steps 1-4 are the same as using `CreateKernelFromBinary`):
1. Run the program normally to generate the kernel binaries
2. Copy the generated kernel directories under the root directory (e.g. from ~/.cache/tt-metal-cache/<hash>/<build_key>/kernels/<kernel_dirs>). You can copy only the directories & binaries you want to use, but you must preserve the full path under the root.
3. Call `tt::tt_metal::experimental::ComputeKernelOriginalPathHash` on the path passed into `CreateKernel` to generate the hash of the original path.
4. Update program to call `tt_metal::experimental::SetKernelBinaryPathPrefix`. Argument `binary_path_prefix` is the location the kernel directories were copied to in step 2.
5. In the `KernelDescriptor`:
	1. `kernel_source` is the name of the kernel
	2. `source_type = tt::tt_metal::KernelDescriptor::SourceType::EXPERIMENTAL_BINARY_PATH`
	3. `binary_hash` is the hash generated by step 3

**Comment by ihnpark551** [comment]
Hi @kstevensTT , thanks for your sharing. We would see the PR and description you shared!

**Comment by ihnpark551** [comment]
@kstevensTT @pgkeller We've checked and concluded that those TT team shared us are all we need, so that this ticket would be closed.
Huge appreciation for your support!

**Comment by pgkeller** [comment]
absolutely.  as discussed before, this solution is a bit cobbled together, we plan to clean this up w/ the new build flow in the coming months.  the general flow should be similar, might vary in some details, tbd

**Comment by ihnpark551** [comment]
Understood, then we'd better to revive this ticket till end of Nov., for clean up work!

**Comment by kstevensTT** [comment]
I am submitting a commit that changes which directory you need to copy and provide to `tt::tt_metal::experimental::ComputeKernelOriginalPathHash`. The change is from:

~/.cache/tt-metal-cache/<hash>/<build_key>

to:

~/.cache/tt-metal-cache/<hash>/<build_key>/kernels

With this commit, you would provide `tt::tt_metal::experimental::ComputeKernelOriginalPathHash` with the `kernels` directory.

The reason for this is because, prior to this commit, the binary was not necessarily portable to machines with different boards, even when those boards have the same architecture. With this new commit and pointing `tt::tt_metal::experimental::ComputeKernelOriginalPathHash` to the `kernels` directory, the binary becomes portable to other machines with the same architecture.

I have updated the original comment to reflect this change.

**Comment by ihnpark551** [comment]
@kstevensTT Thanks for your sharing!
Could I regard that commit as the final version, which means the cleanup version that @pgkeller said?

**Comment by pgkeller** [comment]
> [@kstevensTT](https://github.com/kstevensTT) Thanks for your sharing! Could I regard that commit as the final version, which means the cleanup version that [@pgkeller](https://github.com/pgkeller) said?

no, sorry, cleanup is still coming and is 1-2 months away from completion

**Comment by ihnpark551** [comment]
Oh I see, but thanks for all your progress sharing!

**Comment by ihnpark551** [comment]
Hi @pgkeller , hope you had a warm weekend. Can I have any updates for this ticket?

**Comment by pgkeller** [comment]
Hi @ihnpark551 we are progressing on the refactoring of the build, however the work is paused for some other interrupts and the holidays.  We'll pick it up in the new year, I don't expect we'll have anything for you before the end of January

**Comment by ihnpark551** [comment]
Hi @pgkeller, then could I understand this ticket's due as end of 2026.01?
If it is, then we would use workaround your team provided till then.

**Comment by ihnpark551** [comment]
Hi @abhullar-tt , thanks for your status sharing.
Then, could I understand that when API is fully prepared till mid March, that would be the final output of this ticket?
Or is there any other components more than API, which should be prepared to complete this ticket?

**Comment by abhullar-tt** [comment]
@ihnpark551 there shouldn't be any more components that would be needed for offline compilation. As discussed earlier this year, there would be follow up work to implement kernel src encryption but that can be tracked separately

**Comment by ihnpark551** [comment]
Understood, thanks for your clarification.

**Comment by zzigler-tt** [comment]
@abhullar-tt I know there are other priorities in flight, can you provide an updated timeline if you have it, or will this be a best effort item when we are able to allocate time?

**Comment by abhullar-tt** [comment]
Thanks for reminder ping @zzigler-tt. Much of the offline support was paused to focus on distributed JIT compile speedups. @ruizhangTT will be context switching between some compile optimizations and offline compile support with the optimizations taking precedence. Due to this, we think the offline compile won't be fully available until closer to end of May

**Comment by ruizhangTT** [comment]
https://github.com/tenstorrent/tt-metal/pull/39884 has been merged to main.

This is an API to load precompiled kernels.

The flow to use it is as follows.

1. Run the program normally to generate the kernel binaries
2. Copy the generated kernel directories under the root directory (e.g. from `~/.cache/tt-metal-cache/<build_key>/kernels/<kernel_name>`). The destination must preserve the directory structure under .../kernels:
   - For example, you can copy the entire subdirectory for `my_kernel` by `cp -r ~/.cache/tt-metal-cache/123456/kernels/my_kernel/ /path/to/precompiled/`.
3. Call [tt:tt_metal::experimental::CreateKernelFromPrecompiled](https://github.com/tenstorrent/tt-metal/blob/70277d3a7c828af002e51936fe255b303beaa25f/tt_metal/api/tt-metalium/experimental/offline_kernel_compile.hpp#L88).
   - `file_name` must be the original path used to compile the kernel.
   - In precompiled_config, specify `precompiled_dir` (in the example above, you would specify `/path/to/precompiled`) and `fallback_policy`.
      - `JitCompile` means when it cannot find an appropriate precompiled binary, it will fallback to JIT build.
      - `Error` means it will throw an exception if it cannot find an appropriate precompiled binary.  In this mode, the source code indicated by `file_name` does not need to exist on the host.


Compared to [previous implementation](https://github.com/tenstorrent/tt-metal/issues/11006#issuecomment-3367140708), this flow does not require you to track the hash of the kernel - it will compute the hash at run time and locate the binary that corresponds to the hash in the precompiled directory.

Please be advised that the hash could change in some circumstances:
- Switched to a different product (e.g., WH -> BH).
- Changes in defines or compile-time arguments.
- Changes in run-time options (e.g., watcher, dprint).
- Change in tt-metal version.  (Hash is not guaranteed to be stable across tt-metal versions.)

If hash changes and the new hash does not have a corresponding precompiled binary, the call could fail.  In general, this is safer than manually specifying the hash because the latter could silently load a wrong binary.

---
What's next:

A dedicated API for kernel offline compilation.  This will replace step 1 in the above - you will not need to run the program normally to generate kernel binaries, and you will be able to generate kernel binaries for multiple product with a single API call.

**Comment by ihnpark551** [comment]
Thanks for your sharing @ruizhangTT , I’m happy to wait until offline compilation is ready.

**Comment by zzigler-tt** [comment]
@ruizhangTT what's the current status of offline compilation?
@pgkeller for visibility as this item is being surfaced once again.

**Comment by ruizhangTT** [comment]
The compilation API is pending approval: https://github.com/tenstorrent/tt-metal/pull/43851

This will close the ticket.  In the meantime, it is possible to load kernel binaries that were jit compiled, as indicated by [the comment above](https://github.com/tenstorrent/tt-metal/issues/11006#issuecomment-4306370316)

**Comment by ruizhangTT** [comment]
https://github.com/tenstorrent/tt-metal/pull/43851 has been merged to main.

New API allows for compiling a specific kernel into binary form.  The API signature and documentation can be found at [here](https://github.com/tenstorrent/tt-metal/blob/b8c0bde1623b45f3c0eb6dbbbd75c7a83b29b33a/tt_metal/api/tt-metalium/experimental/offline_kernel_compile.hpp#L140-L173)

There is no remaining work for this issue.

**Comment by ruizhangTT** [comment]
Moreh: feature is complete, can you confirm the new API fulfills your need?  Planning to close this issue.

**Comment by ruizhangTT** [comment]
Closing as the feature is complete.

**Comment by razorback3** [comment]
Hi @ruizhangTT, I found a build-key coverage issue in the offline compilation
API on T3000 fast dispatch.

### Reproducer

1. Compile the kernels used by `ttnn.experimental.rotate_half` with
   `CompileKernelOffline(AllSupportedProducts)`.
2. Make the original kernel sources unavailable.
3. Use a fresh cache and `FallbackPolicy::Error`.
4. Run `rotate_half` on T3000 with fast dispatch.

The loader fails because the runtime-requested artifact was not generated:

    Kernel: bcast_hw
    compile_hash: 0xb89a822de96e6b63
    PrecompiledKernelNotFoundError

### Cause

The live path in `create_jit_device_config()` initializes:

    resolved_dispatch_core_type =
        resolve_dispatch_core_type(...)

However, the offline path in `enumerate_jit_device_configs()` initializes
`dispatch_core_type` but not `resolved_dispatch_core_type`, leaving it at its
default value of `WORKER`.

For T3000, TTNN uses Ethernet dispatch by default, resulting in:

    live runtime:  dispatch_core_type=ETH, resolved_dispatch_core_type=ETH
    offline build: dispatch_core_type=ETH, resolved_dispatch_core_type=WORKER

Since `resolved_dispatch_core_type` is part of the build key, the runtime cannot
find the offline-generated artifact.

As a diagnostic, I initialized the missing field in the offline path and
regenerated the artifacts. The source-less test then passed without JIT
compilation.

Could this issue be reopened to track the fix and a fast-dispatch regression
test, or would you prefer a separate bug?

**Comment by pgkeller** [comment]
@razorback3, please file a separate issue for the bug and assign @ruizhangTT , we'll work on a fix asap.  thanks
